File: backend/app/api/v1/websites.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from uuid import UUID
import re

from app.core.database import get_db
from app.core.deps import get_current_user
from app.domains.users.models import User
from app.domains.businesses.models import Business
from app.domains.websites.models import Website, Domain, WebsiteStatus
from app.domains.websites.schemas import (
    WebsiteCreate, WebsiteUpdate, WebsiteResponse,
    DomainCreate, DomainUpdate, DomainResponse,
    OnboardingCompleteRequest
)

router = APIRouter()


async def _get_business_for_user(
    business_id: UUID, user: User, db: AsyncSession
) -> Business:
    result = await db.execute(
        select(Business).where(
            Business.id == business_id,
            Business.user_id == user.id,
            Business.is_active == True,
        )
    )
    business = result.scalar_one_or_none()
    if not business:
        raise HTTPException(status_code=404, detail="Negocio no encontrado")
    return business


def _validate_subdomain(subdomain: str) -> bool:
    """Validate subdomain format (alphanumeric + hyphens, 3-63 chars)."""
    pattern = r'^[a-z0-9]([a-z0-9-]{1,61}[a-z0-9])?$'
    return bool(re.match(pattern, subdomain.lower()))


@router.post("/onboarding/complete", response_model=WebsiteResponse)
async def complete_onboarding(
    req: OnboardingCompleteRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Complete onboarding: create business + website + domain in one call."""

    # Validate subdomain
    if not _validate_subdomain(req.subdomain):
        raise HTTPException(status_code=400, detail="Subdominio inválido. Solo letras, números y guiones.")

    # Check if subdomain already taken
    domain_check = await db.execute(
        select(Domain).where(Domain.subdomain == req.subdomain.lower())
    )
    if domain_check.scalar_one_or_none():
        raise HTTPException(status_code=400, detail="Subdominio no disponible")

    # Get or create business
    business_query = await db.execute(
        select(Business).where(
            Business.user_id == current_user.id,
            Business.is_active == True,
        )
    )
    business = business_query.scalars().first()

    if not business:
        raise HTTPException(status_code=404, detail="Negocio no encontrado. Crea uno primero.")

    # Update business info if provided
    if req.business_name:
        business.name = req.business_name
    if req.business_description:
        business.description = req.business_description

    # Create website
    website = Website(
        business_id=business.id,
        template=req.website_template,
        title=req.business_name,
        description=req.business_description,
        status=WebsiteStatus.DRAFT,
    )
    db.add(website)
    await db.flush()

    # Create domain
    domain = Domain(
        website_id=website.id,
        subdomain=req.subdomain.lower(),
        is_primary=True,
        is_verified=True,  # Sellia-hosted subdomain always verified
    )
    db.add(domain)

    await db.commit()
    await db.refresh(website)
    await db.refresh(domain)

    website.domain = domain
    return website


@router.post("/businesses/{business_id}/website", response_model=WebsiteResponse, status_code=status.HTTP_201_CREATED)
async def create_website(
    business_id: UUID,
    req: WebsiteCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Create website for a business."""
    await _get_business_for_user(business_id, current_user, db)

    # Validate subdomain
    if not _validate_subdomain(req.subdomain):
        raise HTTPException(status_code=400, detail="Subdominio inválido")

    # Check if subdomain taken
    domain_check = await db.execute(
        select(Domain).where(Domain.subdomain == req.subdomain.lower())
    )
    if domain_check.scalar_one_or_none():
        raise HTTPException(status_code=400, detail="Subdominio no disponible")

    # Check if business already has website
    website_check = await db.execute(
        select(Website).where(Website.business_id == business_id)
    )
    if website_check.scalar_one_or_none():
        raise HTTPException(status_code=400, detail="Negocio ya tiene website")

    website = Website(
        business_id=business_id,
        template=req.template,
        title=req.title,
        description=req.description,
        hero_image_url=req.hero_image_url,
        meta_title=req.meta_title,
        meta_description=req.meta_description,
        meta_keywords=req.meta_keywords,
    )
    db.add(website)
    await db.flush()

    domain = Domain(
        website_id=website.id,
        subdomain=req.subdomain.lower(),
        is_primary=True,
        is_verified=True,
    )
    db.add(domain)

    await db.commit()
    await db.refresh(website)
    await db.refresh(domain)

    website.domain = domain
    return website


@router.get("/businesses/{business_id}/website", response_model=WebsiteResponse)
async def get_website(
    business_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get website for a business."""
    await _get_business_for_user(business_id, current_user, db)

    result = await db.execute(
        select(Website).where(Website.business_id == business_id)
    )
    website = result.scalar_one_or_none()
    if not website:
        raise HTTPException(status_code=404, detail="Website no encontrado")

    return website


@router.put("/businesses/{business_id}/website", response_model=WebsiteResponse)
async def update_website(
    business_id: UUID,
    req: WebsiteUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Update website."""
    await _get_business_for_user(business_id, current_user, db)

    result = await db.execute(
        select(Website).where(Website.business_id == business_id)
    )
    website = result.scalar_one_or_none()
    if not website:
        raise HTTPException(status_code=404, detail="Website no encontrado")

    update_data = req.model_dump(exclude_unset=True)
    if "config" in update_data and website.config:
        update_data["config"] = {**website.config, **update_data["config"]}
    if "pages" in update_data and website.pages:
        update_data["pages"] = {**website.pages, **update_data["pages"]}

    for field, value in update_data.items():
        setattr(website, field, value)

    if req.status == WebsiteStatus.PUBLISHED:
        from datetime import datetime, timezone
        website.published_at = datetime.now(timezone.utc)

    await db.commit()
    await db.refresh(website)
    return website


@router.post("/businesses/{business_id}/website/publish", response_model=WebsiteResponse)
async def publish_website(
    business_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Publish website (make it live)."""
    await _get_business_for_user(business_id, current_user, db)

    result = await db.execute(
        select(Website).where(Website.business_id == business_id)
    )
    website = result.scalar_one_or_none()
    if not website:
        raise HTTPException(status_code=404, detail="Website no encontrado")

    from datetime import datetime, timezone
    website.status = WebsiteStatus.PUBLISHED
    website.published_at = datetime.now(timezone.utc)

    await db.commit()
    await db.refresh(website)
    return website


@router.get("/businesses/{business_id}/domain", response_model=DomainResponse)
async def get_domain(
    business_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get domain info for business website."""
    await _get_business_for_user(business_id, current_user, db)

    result = await db.execute(
        select(Website).where(Website.business_id == business_id)
    )
    website = result.scalar_one_or_none()
    if not website:
        raise HTTPException(status_code=404, detail="Website no encontrado")

    domain_result = await db.execute(
        select(Domain).where(Domain.website_id == website.id)
    )
    domain = domain_result.scalar_one_or_none()
    if not domain:
        raise HTTPException(status_code=404, detail="Dominio no encontrado")

    return domain


@router.put("/businesses/{business_id}/domain", response_model=DomainResponse)
async def update_domain(
    business_id: UUID,
    req: DomainUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Update domain (e.g., add custom domain)."""
    await _get_business_for_user(business_id, current_user, db)

    result = await db.execute(
        select(Website).where(Website.business_id == business_id)
    )
    website = result.scalar_one_or_none()
    if not website:
        raise HTTPException(status_code=404, detail="Website no encontrado")

    domain_result = await db.execute(
        select(Domain).where(Domain.website_id == website.id)
    )
    domain = domain_result.scalar_one_or_none()
    if not domain:
        raise HTTPException(status_code=404, detail="Dominio no encontrado")

    if req.custom_domain:
        # Check if custom domain already taken
        custom_check = await db.execute(
            select(Domain).where(
                Domain.custom_domain == req.custom_domain,
                Domain.id != domain.id
            )
        )
        if custom_check.scalar_one_or_none():
            raise HTTPException(status_code=400, detail="Dominio personalizado ya en uso")

        domain.custom_domain = req.custom_domain
        domain.is_verified = False  # Needs DNS verification

    if req.is_verified is not None:
        domain.is_verified = req.is_verified

    await db.commit()
    await db.refresh(domain)
    return domain

File: frontend/src/lib/api/websites.ts
/**
 * Website API client — Tier 1 onboarding + website builder
 */

export interface Website {
  id: string
  business_id: string
  template: 'MINIMAL' | 'CLASSIC' | 'MODERN' | 'ECOMMERCE' | 'PORTFOLIO'
  status: 'DRAFT' | 'PUBLISHED' | 'ARCHIVED'
  title?: string
  description?: string
  hero_image_url?: string
  meta_title?: string
  meta_description?: string
  meta_keywords?: string
  og_image_url?: string
  config: Record<string, any>
  pages: Record<string, any>
  domain?: Domain
  created_at: string
  updated_at: string
  published_at?: string
}

export interface Domain {
  id: string
  website_id: string
  subdomain: string
  custom_domain?: string
  is_primary: boolean
  is_verified: boolean
  dns_records: Record<string, any>
  created_at: string
  updated_at: string
}

export interface OnboardingRequest {
  business_name: string
  business_description: string
  website_template: string
  subdomain: string
}

const API_BASE = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000'

export const websites = {
  /**
   * Complete onboarding flow (create business + website + domain)
   */
  completeOnboarding: async (data: OnboardingRequest): Promise<Website> => {
    const res = await fetch(`${API_BASE}/api/v1/websites/onboarding/complete`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      credentials: 'include',
      body: JSON.stringify(data),
    })

    if (!res.ok) {
      const err = await res.json()
      throw new Error(err.detail || 'Error en onboarding')
    }

    return res.json()
  },

  /**
   * Get website for a business
   */
  getWebsite: async (businessId: string): Promise<Website> => {
    const res = await fetch(`${API_BASE}/api/v1/websites/businesses/${businessId}/website`, {
      credentials: 'include',
    })

    if (!res.ok) {
      throw new Error('Error obteniendo website')
    }

    return res.json()
  },

  /**
   * Create website for a business
   */
  createWebsite: async (
    businessId: string,
    data: {
      template: string
      title?: string
      description?: string
      subdomain: string
    }
  ): Promise<Website> => {
    const res = await fetch(`${API_BASE}/api/v1/websites/businesses/${businessId}/website`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      credentials: 'include',
      body: JSON.stringify(data),
    })

    if (!res.ok) {
      const err = await res.json()
      throw new Error(err.detail || 'Error creando website')
    }

    return res.json()
  },

  /**
   * Update website
   */
  updateWebsite: async (
    businessId: string,
    data: Partial<Website>
  ): Promise<Website> => {
    const res = await fetch(`${API_BASE}/api/v1/websites/businesses/${businessId}/website`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      credentials: 'include',
      body: JSON.stringify(data),
    })

    if (!res.ok) {
      throw new Error('Error actualizando website')
    }

    return res.json()
  },

  /**
   * Publish website
   */
  publishWebsite: async (businessId: string): Promise<Website> => {
    const res = await fetch(`${API_BASE}/api/v1/websites/businesses/${businessId}/website/publish`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      credentials: 'include',
    })

    if (!res.ok) {
      throw new Error('Error publicando website')
    }

    return res.json()
  },

  /**
   * Get domain info
   */
  getDomain: async (businessId: string): Promise<Domain> => {
    const res = await fetch(`${API_BASE}/api/v1/websites/businesses/${businessId}/domain`, {
      credentials: 'include',
    })

    if (!res.ok) {
      throw new Error('Error obteniendo dominio')
    }

    return res.json()
  },

  /**
   * Update domain
   */
  updateDomain: async (
    businessId: string,
    data: { custom_domain?: string; is_verified?: boolean }
  ): Promise<Domain> => {
    const res = await fetch(`${API_BASE}/api/v1/websites/businesses/${businessId}/domain`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      credentials: 'include',
      body: JSON.stringify(data),
    })

    if (!res.ok) {
      throw new Error('Error actualizando dominio')
    }

    return res.json()
  },
}
